### PR TITLE
fix: set dhcp gateway to U2OInterconnectionIP when enabling dhcp and u2o

### LINKED ISSUE #4224: [BUG] 开启U2O和DHCP，虚机通过DHCP获取到的网关仍是原网关
### Kube-OVN Version

v1.13.0

### Kubernetes Version

v1.25.3

### Operation-system/Kernel Version

"Ubuntu 22.04.4 LTS" 6.5.0-41-generic

### Description

image
开启u2o后，虚机通过dhcp获取到的gateway还是subnet指定的gateway，而不是lrp的地址

### Steps To Reproduce

1、开启子网dhcp和u2o
2、创建虚机通过dhcp获取默认路由

### Current Behavior

虚机通过dhcp获取到的默认路由是子网指定的gateway而不是u2o开启后的lrp地址

### Expected Behavior

开启子网dhcp和u2o后虚机通过dhcp获取到lrp的地址作为默认路由的网关

### PR DESCRIPTION
**Pull request by zcq98**
# Pull Request

- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Bug fixes

## Which issue(s) this PR fixes

Fixes #4224

### CODE REVIEW

File: pkg/ovs/ovn-nb-dhcp_options.go [hunk 1]
```diff
@@ -69,6 +69,10 @@ func (c *OVNNbClient) UpdateDHCPOptions(subnet *kubeovnv1.Subnet, mtu int) (*DHC
 		v4Gateway = gateways[0]
 	}
 
+	if subnet.Spec.U2OInterconnection {
+		v4Gateway = subnet.Status.U2OInterconnectionIP
```
Comment (zhangzujian) [inline pkg/ovs/ovn-nb-dhcp_options.go:73]: We should consider dual-stack subnets.

**Comment by zhangzujian** [comment]
We need an e2e test case.

File: test/e2e/kube-ovn/underlay/underlay.go [hunk 2]
```diff
@@ -919,6 +922,16 @@ func checkU2OItems(f *framework.Framework, subnet *apiv1.Subnet, underlayPod, ov
 				framework.ExpectEqual(subnet.Spec.U2OInterconnectionIP, subnet.Status.U2OInterconnectionIP)
 			}
 		}
+		if f.HasIPv4() && subnet.Spec.EnableDHCP {
+			if !f.VersionPriorTo(1, 12) {
+				ginkgo.By("checking subnet's u2o dhcp gateway ip of underlay subnet " + subnet.Name)
```
Comment (bobz965) [inline test/e2e/kube-ovn/underlay/underlay.go:927]:
```suggestion
				ginkgo.By("checking u2o dhcp gateway ip of underlay subnet " + subnet.Name)
```

File: test/e2e/kube-ovn/underlay/underlay.go [hunk 3]
```diff
@@ -567,6 +567,9 @@ var _ = framework.SerialDescribe("[group:underlay]", func() {
 		ginkgo.By("Creating underlay subnet " + subnetName)
 		subnet := framework.MakeSubnet(subnetName, vlanName, strings.Join(underlayCidr, ","), strings.Join(gateway, ","), "", "", excludeIPs, nil, []string{namespaceName})
 		subnet.Spec.U2OInterconnection = true
+		if f.HasIPv4() {
```
Comment (bobz965) [inline test/e2e/kube-ovn/underlay/underlay.go:570]: 加个注释，说明只有 IPv4 需要这个功能

**Review by bobz965 — APPROVED** [review]

File: test/e2e/kube-ovn/underlay/underlay.go [hunk 4]
```diff
@@ -567,6 +567,10 @@ var _ = framework.SerialDescribe("[group:underlay]", func() {
 		ginkgo.By("Creating underlay subnet " + subnetName)
 		subnet := framework.MakeSubnet(subnetName, vlanName, strings.Join(underlayCidr, ","), strings.Join(gateway, ","), "", "", excludeIPs, nil, []string{namespaceName})
 		subnet.Spec.U2OInterconnection = true
+		// only ipv4 needs to verify that the gateway address is consistent with U2OInterconnectionIP when enabling DHCP and U20
```
Comment (zhangzujian) [inline test/e2e/kube-ovn/underlay/underlay.go:570]:
```suggestion
		// only ipv4 needs to verify that the gateway address is consistent with U2OInterconnectionIP when enabling DHCP and U2O
```

**Review by zhangzujian — APPROVED** [review]

**Pull request merged**